# I- Installation des dépendances

In [2]:
!dir

 Le volume dans le lecteur C n'a pas de nom.
 Le num‚ro de s‚rie du volume est E8AF-AAB4

 R‚pertoire de C:\Users\HP\Documents\IGIT3\NLP\TPs\Translator

03/03/2025  10:28 AM    <DIR>          .
03/03/2025  10:28 AM    <DIR>          ..
02/21/2025  02:53 PM    <DIR>          .ipynb_checkpoints
03/03/2025  09:53 AM             3,316 app.py
02/27/2025  12:04 PM    <DIR>          archives
03/02/2025  11:54 PM                 0 dummy_output.txt
02/27/2025  12:14 PM    <DIR>          fr-en_data
03/03/2025  10:27 AM           458,396 Translator_Notebook.ipynb
               3 fichier(s)          461,712 octets
               5 R‚p(s)  242,836,123,648 octets libres


In [3]:
! python --version

Python 3.11.4


In [ ]:
!pip3 install --upgrade pip

In [4]:
!pip install OpenNMT-py

   ---------------------------------------- 0.0/262.8 kB ? eta -:--:--
   ----------------------------- ---------- 194.6/262.8 kB ? eta -:--:--
   -------------------------------------- - 256.0/262.8 kB 4.0 MB/s eta 0:00:01
   -------------------------------------- - 256.0/262.8 kB 4.0 MB/s eta 0:00:01
   -------------------------------------- - 256.0/262.8 kB 4.0 MB/s eta 0:00:01
   -------------------------------------- 262.8/262.8 kB 172.0 kB/s eta 0:00:00
   ---------------------------------------- 0.0/19.5 MB ? eta -:--:--
    --------------------------------------- 0.3/19.5 MB 5.4 MB/s eta 0:00:04
    --------------------------------------- 0.4/19.5 MB 5.4 MB/s eta 0:00:04
   - -------------------------------------- 0.6/19.5 MB 4.8 MB/s eta 0:00:04
   - -------------------------------------- 0.8/19.5 MB 4.5 MB/s eta 0:00:05
   -- ------------------------------------- 1.1/19.5 MB 5.5 MB/s eta 0:00:04
   --- ------------------------------------ 1.5/19.5 MB 5.7 MB/s eta 0:00:04
   -

ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
opennmt-tf 2.32.0 requires ctranslate2<4,>=3.0, but you have ctranslate2 4.5.0 which is incompatible.


In [5]:
!pip uninstall -y torch torchvision torchaudio

Found existing installation: torch 2.1.0
Uninstalling torch-2.1.0:
  Successfully uninstalled torch-2.1.0


In [8]:
!pip install torch==2.1.0 torchvision==0.16.0 torchaudio==2.1.0

   ---------------------------------------- 0.0/192.3 MB ? eta -:--:--
   ---------------------------------------- 0.2/192.3 MB 4.6 MB/s eta 0:00:42
   ---------------------------------------- 0.5/192.3 MB 6.7 MB/s eta 0:00:29
   ---------------------------------------- 1.0/192.3 MB 8.3 MB/s eta 0:00:24
   ---------------------------------------- 1.0/192.3 MB 8.3 MB/s eta 0:00:24
   ---------------------------------------- 1.0/192.3 MB 8.3 MB/s eta 0:00:24
   ---------------------------------------- 1.1/192.3 MB 4.1 MB/s eta 0:00:47
   ---------------------------------------- 1.7/192.3 MB 5.2 MB/s eta 0:00:37
   ---------------------------------------- 2.2/192.3 MB 5.9 MB/s eta 0:00:33
    --------------------------------------- 2.5/192.3 MB 6.0 MB/s eta 0:00:32
    --------------------------------------- 2.6/192.3 MB 6.1 MB/s eta 0:00:32
    --------------------------------------- 2.6/192.3 MB 6.1 MB/s eta 0:00:32
    --------------------------------------- 2.6/192.3 MB 6.1 MB/s eta 0

In [8]:
!pip list --format=freeze

absl-py==1.4.0
accelerate==1.3.0
aiohappyeyeballs==2.4.6
aiohttp==3.11.12
aiosignal==1.3.2
alabaster==1.0.0
albucore==0.0.23
albumentations==2.0.4
ale-py==0.10.2
altair==5.5.0
annotated-types==0.7.0
anyio==3.7.1
argon2-cffi==23.1.0
argon2-cffi-bindings==21.2.0
array_record==0.6.0
arviz==0.20.0
astropy==7.0.1
astropy-iers-data==0.2025.2.17.0.34.13
astunparse==1.6.3
atpublic==4.1.0
attrs==25.1.0
audioread==3.0.1
autograd==1.7.0
babel==2.17.0
backcall==0.2.0
beautifulsoup4==4.13.3
betterproto==2.0.0b6
bigframes==1.37.0
bigquery-magics==0.6.0
bleach==6.2.0
blinker==1.9.0
blis==0.7.11
blosc2==3.1.1
bokeh==3.6.3
Bottleneck==1.4.2
bqplot==0.12.44
branca==0.8.1
CacheControl==0.14.2
cachetools==5.5.1
catalogue==2.0.10
certifi==2025.1.31
cffi==1.17.1
chardet==5.2.0
charset-normalizer==3.4.1
chex==0.1.88
clarabel==0.10.0
click==8.1.8
cloudpathlib==0.20.0
cloudpickle==3.1.1
cmake==3.31.4
cmdstanpy==1.2.5
colorama==0.4.6
colorcet==3.1.0
colorlover==0.3.0
colour==0.1.5
community==1.0.0b1
confection=

In [3]:
# checking for well installing
!onmt_train --help

usage: onmt_train [-h] [-config CONFIG] [-save_config SAVE_CONFIG] -data DATA
                  [-skip_empty_level {silent,warning,error}]
                  [-transforms {clean,normalize,switchout,tokendrop,tokenmask,inlinetags,filtertoolong,prefix,suffix,sentencepiece,bpe,onmt_tokenize,uppercase,insert_mask_before_placeholder,terminology,docify,inferfeats,bart,fuzzymatch} [{clean,normalize,switchout,tokendrop,tokenmask,inlinetags,filtertoolong,prefix,suffix,sentencepiece,bpe,onmt_tokenize,uppercase,insert_mask_before_placeholder,terminology,docify,inferfeats,bart,fuzzymatch} ...]]
                  [-save_data SAVE_DATA] [-overwrite] [-n_sample N_SAMPLE] [-dump_transforms]
                  -src_vocab SRC_VOCAB [-tgt_vocab TGT_VOCAB] [-share_vocab]
                  [--decoder_start_token DECODER_START_TOKEN]
                  [--default_specials DEFAULT_SPECIALS [DEFAULT_SPECIALS ...]]
                  [-n_src_feats N_SRC_FEATS] [-src_feats_defaults SRC_FEATS_DEFAULTS]
             

In [ ]:
# Data provenance
# Europarl : Un corpus de textes du Parlement européen.
# https://www.statmt.org/europarl/


# II- Loading and exploring dataset

In [1]:
%cd fr-en_data
!dir

C:\Users\HP\Documents\IGIT3\NLP\TPs\Translator\fr-en_data
 Le volume dans le lecteur C n'a pas de nom.
 Le num‚ro de s‚rie du volume est E8AF-AAB4

 R‚pertoire de C:\Users\HP\Documents\IGIT3\NLP\TPs\Translator\fr-en_data

02/27/2025  12:14 PM    <DIR>          .
02/27/2025  12:14 PM    <DIR>          ..
02/24/2025  08:18 AM    <DIR>          .ipynb_checkpoints
03/09/2025  02:24 PM             1,150 config.yaml
03/03/2025  10:25 AM    <DIR>          data
11/21/2011  03:50 PM       301,523,301 europarl-v7.fr-en.en
02/27/2025  12:13 PM       346,919,801 europarl-v7.fr-en.fr
03/02/2025  11:07 PM    <DIR>          models
02/23/2025  06:14 PM             8,063 output.log
               4 fichier(s)      648,452,315 octets
               5 R‚p(s)  243,911,909,376 octets libres


In [13]:
# Source Langage
with open("europarl-v7.fr-en.fr", "r", encoding="utf-8") as file:
  src_lines = file.readlines()

In [14]:
# Target Langage
with open("europarl-v7.fr-en.en", "r", encoding="utf-8") as file:
  tgt_lines = file.readlines()

In [15]:
print(f"data source length: {len(src_lines)}")
print(f"data target length: {len(tgt_lines)}")

data source length: 2007723
data target length: 2007723


In [16]:
for i in range(10):
    print(f"FR: {src_lines[i]}")
    print(f"EN: {tgt_lines[i]}")
    print("-" * 50)

FR: Reprise de la session

EN: Resumption of the session

--------------------------------------------------
FR: Je déclare reprise la session du Parlement européen qui avait été interrompue le vendredi 17 décembre dernier et je vous renouvelle tous mes vux en espérant que vous avez passé de bonnes vacances.

EN: I declare resumed the session of the European Parliament adjourned on Friday 17 December 1999, and I would like once again to wish you a happy new year in the hope that you enjoyed a pleasant festive period.

--------------------------------------------------
FR: Comme vous avez pu le constater, le grand "bogue de l'an 2000" ne s'est pas produit. En revanche, les citoyens d'un certain nombre de nos pays ont été victimes de catastrophes naturelles qui ont vraiment été terribles.

EN: Although, as you will have seen, the dreaded 'millennium bug' failed to materialise, still the people in a number of countries suffered a series of natural disasters that truly were dreadful.

----

# III- Cleaning dataset

In [ ]:
def clean_text(text):
    text = text.lower()
    text = text.strip()
    text = " ".join(text.split())
    return text


In [ ]:
src_clean = [clean_text(line) for line in src_lines]
tgt_clean = [clean_text(line) for line in tgt_lines]
for i in range(5):
    print(f"FR: {src_clean[i]}")
    print(f"EN: {tgt_clean[i]}")
    print("-" * 50)

FR: reprise de la session
EN: resumption of the session
--------------------------------------------------
FR: je déclare reprise la session du parlement européen qui avait été interrompue le vendredi 17 décembre dernier et je vous renouvelle tous mes vux en espérant que vous avez passé de bonnes vacances.
EN: i declare resumed the session of the european parliament adjourned on friday 17 december 1999, and i would like once again to wish you a happy new year in the hope that you enjoyed a pleasant festive period.
--------------------------------------------------
FR: comme vous avez pu le constater, le grand "bogue de l'an 2000" ne s'est pas produit. en revanche, les citoyens d'un certain nombre de nos pays ont été victimes de catastrophes naturelles qui ont vraiment été terribles.
EN: although, as you will have seen, the dreaded 'millennium bug' failed to materialise, still the people in a number of countries suffered a series of natural disasters that truly were dreadful.
----------

# IV- Saving cleaned data

## IV.1 Splitting data to train, test and validation

In [ ]:
from sklearn.model_selection import train_test_split

src_train, src_temp, tgt_train, tgt_temp = train_test_split(src_clean, tgt_clean, test_size=0.2, random_state=42)
src_val, src_test, tgt_val, tgt_test = train_test_split(src_temp, tgt_temp, test_size=0.5, random_state=42)


In [ ]:
print(f"source train data length:{len(src_train)}")
print(f"source test data length:{len(src_test)}")
print(f"source validation data length:{len(src_val)}")
print(f"target train data length:{len(tgt_train)}")
print(f"target test data length:{len(tgt_test)}")
print(f"target validation data length:{len(tgt_val)}")

source train data length:1606178
source test data length:200773
source validation data length:200772
target train data length:1606178
target test data length:200773
target validation data length:200772


## IV.2 Saving

In [ ]:
def save_data(filename, data):
    with open(filename, "w", encoding="utf-8") as f:
        f.writelines(line + "\n" for line in data)


In [2]:
! dir

 Le volume dans le lecteur C n'a pas de nom.
 Le num‚ro de s‚rie du volume est E8AF-AAB4

 R‚pertoire de C:\Users\HP\Documents\IGIT3\NLP\TPs\Translator\fr-en_data

02/27/2025  12:14 PM    <DIR>          .
02/27/2025  12:14 PM    <DIR>          ..
02/24/2025  08:18 AM    <DIR>          .ipynb_checkpoints
03/09/2025  02:24 PM             1,150 config.yaml
03/03/2025  10:25 AM    <DIR>          data
11/21/2011  03:50 PM       301,523,301 europarl-v7.fr-en.en
02/27/2025  12:13 PM       346,919,801 europarl-v7.fr-en.fr
03/02/2025  11:07 PM    <DIR>          models
02/23/2025  06:14 PM             8,063 output.log
               4 fichier(s)      648,452,315 octets
               5 R‚p(s)  243,911,802,880 octets libres


%mkdir ./data

In [ ]:
save_data("./data/fr_train.txt", src_train)
save_data("./data/en_train.txt", tgt_train)
save_data("./data/fr_val.txt", src_val)
save_data("./data/en_val.txt", tgt_val)
save_data("./data/fr_test.txt", src_test)
save_data("./data/en_test.txt", tgt_test)

# V- OpenNMT prepossing data step

## V.1- Tokenization of train and validation data

In [ ]:
# We shall use Spacy, then let load french and english tokenizer
import spacy

In [ ]:
!python -m spacy download fr_core_news_sm
!python -m spacy download en_core_web_sm

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 16.3/16.3 MB 33.8 MB/s eta 0:00:00
✔ Download and installation successful
You can now load the package via spacy.load('fr_core_news_sm')
⚠ Restart to reload dependencies
If you are in a Jupyter or Colab notebook, you may need to restart Python in
order to load all the package's dependencies. You can do this by selecting the
'Restart kernel' or 'Restart runtime' option.
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.8/12.8 MB 33.4 MB/s eta 0:00:00
✔ Download and installation successful
You can now load the package via spacy.load('en_core_web_sm')
⚠ Restart to reload dependencies
If you are in a Jupyter or Colab notebook, you may need to restart Python in
order to load all the package's dependencies. You can do this by selecting the
'Restart kernel' or 'Restart runtime' option.


In [ ]:
nlp_fr = spacy.load("fr_core_news_sm")
nlp_en = spacy.load("en_core_web_sm")

In [ ]:
print("French model :", nlp_fr)
print("English model :", nlp_en)

French model : <spacy.lang.fr.French object at 0x7d78def6ba10>
English model : <spacy.lang.en.English object at 0x7d78dea12650>


In [ ]:
!pip install tqdm # Follow the process

In [ ]:
from tqdm import tqdm

# Tokenize per batch
def tokenize_batch(lines, nlp, batch_size=1000):
    tokenized_lines = []
    for doc in tqdm(nlp.pipe(lines, batch_size=batch_size), total=len(lines), desc="Tokenisation"):
        tokenized_lines.append(" ".join([token.text for token in doc]))
    return tokenized_lines

In [ ]:
src_val = tokenize_batch(src_val, nlp_fr)
tgt_val = tokenize_batch(tgt_val, nlp_en)
src_train = tokenize_batch(src_train, nlp_fr)
tgt_train = tokenize_batch(tgt_train, nlp_en)

Tokenisation: 100%|██████████| 1606178/1606178 [1:32:43<00:00, 288.70it/s]


In [ ]:
def save_data(filename, data):
    with open(filename, "w", encoding="utf-8") as f:
        f.writelines(line + "\n" for line in data)

In [ ]:
save_data("./data/fr_train_tokenized.txt", src_train)
save_data("./data/en_train_tokenized.txt", tgt_train)
save_data("./data/fr_val_tokenized.txt", src_val)
save_data("./data/en_val_tokenized.txt", tgt_val)

## V.2- Vocabulary creating

In [4]:
# Configuation file yaml file (file configuration wich allow vocab creating, models training, validation and testing)
yaml_path = "config.yaml"
yaml_content = """\

data:
  corpus_1:
    weight: 1
    path_src: ./data/fr_train_tokenized.txt
    path_tgt: ./data/en_train_tokenized.txt
    transforms: []
  valid:
    weight: 1
    path_src: ./data/fr_val_tokenized.txt
    path_tgt: ./data/en_val_tokenized.txt
    transforms: []

src_vocab: ./data/processed_vocab/vocab.src
tgt_vocab: ./data/processed_vocab/vocab.tgt
src_vocab_size: 50000
tgt_vocab_size: 50000

save_model: ./models/fr-en-model
save_checkpoint_steps: 4000
valid_steps: 4000
train_steps: 200000

# Batching
batch_size: 4096
batch_type: "tokens"
valid_batch_size: 2048
accum_count: [4]
accum_steps: [0]

# Optimization
model_dtype: "fp16"
optim: "adam"
learning_rate: 2
warmup_steps: 8000
decay_method: "noam"
adam_beta2: 0.998
max_grad_norm: 0
label_smoothing: 0.1

# Model
encoder_type: transformer
decoder_type: transformer
position_encoding: true
enc_layers: 6
dec_layers: 6
heads: 8
hidden_size: 512
word_vec_size: 512
transformer_ff: 2048
dropout_steps: [0]
dropout: [0.1]
attention_dropout: [0.1]

# Multi-GPU (modifie selon ton setup)
world_size: 1
gpu_ranks: [0]
num_workers: 2

"""

# Écriture du fichier YAML
with open(yaml_path, "w") as f:
    f.write(yaml_content)

print(f"Configuration file created succesfully : {yaml_path}")

Configuration file created succesfully : config.yaml


##### Vocabulary creating step

In [14]:
!python -m onmt.bin.build_vocab -config config.yaml -save_data ./data/processed_vocab -n_sample -1

[2025-02-23 15:56:08,929 INFO] Counter vocab from -1 samples.
[2025-02-23 15:56:08,930 INFO] n_sample=-1: Build vocab on full datasets.
[2025-02-23 15:57:21,175 INFO] Counters src: 127749
[2025-02-23 15:57:21,175 INFO] Counters tgt: 94042


### Verifying the vocabulary created

In [15]:
# Reading of vocab.src file
with open("./data/processed_vocab/vocab.src", 'r', encoding='utf-8') as file:
    lines = file.readlines()

# replaccing of tabulation to space
corrected_lines = [line.replace('\t', ' ') for line in lines]

# Saving the right vocab for vocab.src
with open("./data/processed_vocab/vocab.src", 'w', encoding='utf-8') as file:
    file.writelines(corrected_lines)

In [16]:
# Reading of vocab.tgt file
with open("./data/processed_vocab/vocab.tgt", 'r', encoding='utf-8') as file:
    lines = file.readlines()

# Replacing of tabulation to space
corrected_lines = [line.replace('\t', ' ') for line in lines]

# Saving of vocab.tgt file
with open("./data/processed_vocab/vocab.tgt", 'w', encoding='utf-8') as file:
    file.writelines(corrected_lines)

In [17]:
#  Let verify, its seem there is a problem of vocab.tgt
with open("./data/processed_vocab/vocab.tgt", 'r', encoding='utf-8') as file:
  vocab = file.readlines()

vocab

['the 3080212\n',
 ', 2136716\n',
 '. 1575780\n',
 'of 1459322\n',
 'to 1357293\n',
 'and 1157405\n',
 'in 964051\n',
 'that 737731\n',
 'a 681216\n',
 'is 679256\n',
 'for 471115\n',
 'we 466005\n',
 'i 443769\n',
 'this 435892\n',
 'on 415006\n',
 'it 380276\n',
 'be 359588\n',
 '- 333709\n',
 'not 305142\n',
 'are 301855\n',
 'as 288904\n',
 'have 287016\n',
 'with 262386\n',
 'which 246938\n',
 'european 238702\n',
 'will 212414\n',
 'by 212212\n',
 'has 186192\n',
 'mr 157985\n',
 'at 148052\n',
 'commission 142911\n',
 'an 138375\n',
 'would 135917\n',
 'also 134173\n',
 'all 133005\n',
 'should 127920\n',
 'but 127528\n',
 'can 127045\n',
 'our 125502\n',
 'president 125056\n',
 'from 123964\n',
 'must 116929\n',
 'there 109093\n',
 'been 107845\n',
 'you 105613\n',
 'union 105559\n',
 'parliament 101140\n',
 "'s 99536\n",
 "' 96456\n",
 'states 96163\n',
 'member 93822\n',
 'more 91524\n',
 'was 90860\n',
 'report 88912\n',
 'its 88080\n',
 'or 87643\n',
 'these 87081\n',
 'the

In [18]:
#Analysing of vocab.tgt file
with open("./data/processed_vocab/vocab.tgt", 'r', encoding='utf-8') as file:
    lines = file.readlines()

for i, line in enumerate(lines):
    parts = line.strip().split()
    if len(parts) != 2 or not parts[1].isdigit():
        print(f"Ligne {i + 1} mal formatée : {line.strip()}")

Ligne 2285 mal formatée : 1628


It seems line 2285 is wrong formated. Let correct it

In [19]:
# Reading vocab.tgt
with open("./data/processed_vocab/vocab.tgt", 'r', encoding='utf-8') as file:
    lines = file.readlines()

# Filtering
corrected_lines = []
for line in lines:
    parts = line.strip().split()
    if len(parts) == 2 and parts[1].isdigit():  # Vérifier que la ligne est bien formatée
        corrected_lines.append(line)

# Saving
with open("./data/processed_vocab/vocab.tgt", 'w', encoding='utf-8') as file:
    file.writelines(corrected_lines)

# VI. Trainning

### Verifying the presence of GPU

import torch
torch.cuda.is_available()

!nvidia-smi

## Training step it self

!python -m onmt.bin.train -config config.yaml -gpu_ranks 0

In [7]:
!python -m onmt.bin.train -config config.yaml

# VII- Testing

In [3]:
!onmt_translate -model ./models/fr-en-model_step_44000.pt -src ./data/test1.txt -output ./data/en_pred.txt

Time w/o python interpreter load/terminate:  1055.3151395320892


[2025-03-10 14:39:07,339 INFO] Loading checkpoint from ./models/fr-en-model_step_44000.pt
[2025-03-10 14:49:27,849 INFO] Loading data into the model
[2025-03-10 14:56:42,630 INFO] PRED SCORE: -0.5395, PRED PPL: 1.72 NB SENTENCES: 8


# VIII- Evaluation

In [ ]:
!pip install sacrebleu

In [ ]:
!sacrebleu en_test.txt -i en_pred.txt